src/Core/Event.js

```diff
@@ -14,32 +14,42 @@ var Event = {
 
 		this[type].set(DOMNode(element), element)
 	},
+	dispatchEvent: function (element, type, event) {
+		try {
+			var host = element.host
+			var instance = host.instance
+			var callback = element.event[type]
+			var state = null
+
+			if (!callback)
+				return
+
+			if (typeof callback === 'function')
+				state = callback.call(instance, event)
+			else if (typeof callback.handleEvent === 'function')
+				state = callback.handleEvent(event)
+
+			if (instance && state)
+				lifecycleReturn(host, state)
+		} catch (e) {
+			errorBoundary(host, e, type+':'+getDisplayName(callback.handleEvent || callback), 0)
+		}
+	},
 	/**
 	 * @param {Event} event
 	 */
 	handleEvent: function handleEvent (event) {
 		var type = 'on'+event.type
-		var element = this[type].get(event.target)
-
-		if (element)
-			try {
-				var host = element.host
-				var instance = host.instance
-				var callback = element.event[type]
-				var state = null
-
-				if (!callback)
-					return
-
-				if (typeof callback === 'function')
-					state = callback.call(instance, event)
-				else if (typeof callback.handleEvent === 'function')
-					state = callback.handleEvent(event)
+		var target = event.target
+		var bubbles = event.bubbles
+		var map = this[type]
 
-				if (instance && state)
-					lifecycleReturn(host, state)
-			} catch (e) {
-				errorBoundary(host, e, type+':'+getDisplayName(callback.handleEvent || callback), 0)
-			}
+		while (target)
+			if (map.has(target))
+				return this.dispatchEvent(map.get(target), type, defineProperty(event, 'currentTarget', {value: target}))
```

Comment:
bubbling should not stop when handled, only if `e.stopPropagation()` is called.
this also means mutating `currentTarget` is bad because an async event handler has to copy the value

Reply:
By accessing `DOM.parentNode` directly I'm guessing it kills most/all of the performance gain of event delegation. Might as well just let the browser handle it.

Reply:
Yes the recursive `parentNode` access does look bad :(, have to re-eval the cost/benefit, though walking up a large tree is widely cheeper than `addEventListener` –[https://esbench.com/bench/59a4e17f99634800a0348f46](https://esbench.com/bench/59a4e17f99634800a0348f46)